File: Directory Bolt Motia/steps/stream-rbac/00-stream-rbac-message.stream.ts
import type { StreamConfig, StreamSubscription } from 'motia'
import { z } from 'zod'

type RbacStreamAuthContext = { permissions?: string }

export const messageSchema = z.object({
  message: z.string(),
  from: z.enum(['user', 'assistant']),
  status: z.enum(['created', 'pending', 'completed']),
})

export const config: StreamConfig = {
  name: 'rbac_message',
  schema: messageSchema,
  baseConfig: { storageType: 'default' },
  canAccess: (_subscription: StreamSubscription, authContext: RbacStreamAuthContext): boolean | Promise<boolean> => {
    return authContext?.permissions === 'nodejs' || false
  },
}

File: Directory Bolt Motia/steps/stream-rbac/00-stream-rbac-message_stream.py
async def can_access(subscription, auth_context):
    """
    Check if the user has permission to access the RBAC message stream.
    
    Args:
        subscription: The stream subscription details
        auth_context: Authentication context containing user permissions
    
    Returns:
        bool: True if user can access the stream, False otherwise
    """
    return auth_context.get("permissions") == "python" if auth_context else False


config = {
    "name": "rbac_message_python",
    "schema": {
        "type": "object",
        "properties": {
            "message": {"type": "string"},
            "from": {"type": "string", "enum": ["user", "assistant"]},
            "status": {"type": "string", "enum": ["created", "pending", "completed"]},
        },
        "required": ["message", "from", "status"],
    },
    "baseConfig": {"storageType": "default"},
    "canAccess": can_access,
}

